packages/playwright/src/runner/projectUtils.ts

```diff
@@ -19,39 +19,49 @@ import path from 'path';
 import { minimatch } from 'playwright-core/lib/utilsBundle';
 import { promisify } from 'util';
 import type { FullProjectInternal } from '../common/config';
-import { createFileMatcher } from '../util';
+import { createFileMatcher, forceRegExp } from '../util';
 
 const readFileAsync = promisify(fs.readFile);
 const readDirAsync = promisify(fs.readdir);
 
-// The difference to forceRegExp is that we want to match the whole string.
-function forceBoundedRegExp(pattern: string): RegExp {
-  const match = pattern.match(/^\/(.*)\/([gi]*)$/);
-  if (match)
-    return new RegExp(match[1], match[2]);
-  return new RegExp(`^${pattern}$`, 'gi');
-}
-
-export function filterProjects(projects: FullProjectInternal[], projectNames?: string[]): FullProjectInternal[] {
-  if (!projectNames)
+export function filterProjects(projects: FullProjectInternal[], projectNames?: string[], projectGrep?: string): FullProjectInternal[] {
+  if (!projectNames && !projectGrep)
     return [...projects];
-  const unmatchedProjectFilters = new Set<string>(projectNames);
+
+  const names = projects.map(p => p.project.name).filter(name => !!name);
+  if (!names.length)
+    throw new Error(`No named projects are specified in the configuration file`);
+
+  if (projectGrep) {
+    const regex = forceRegExp(projectGrep);
+    const result = projects.filter(project => regex.test(project.project.name));
```

Comment:
When testing on the same RegExp instance, you need to reset `lastIndex` property.